### PR TITLE
Improved sharding algorithms

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/538/

**Comment by graetzer** [comment]
you should add catch tests for these, no ?

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -428,6 +428,7 @@ static int distributeBabyOnShards(
     if (userSpecifiedKey &&
         (!usesDefaultShardingAttributes || !collinfo->allowUserKeys()) &&
         !isRestore) {
+      LOG_TOPIC(ERR, Logger::FIXME) << value.toJson();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]: please remove

File: js/common/tests/shell/shell-community-sharding-cluster.js [hunk 2]
```diff
@@ -139,126 +156,147 @@ function DocumentShardingSuite() {
...
+
+      assertEqual([ 188, 192, 198, 204, 218 ], Object.values(c2.count(true)).sort());
+
+      for (let i = 0; i < 1000; ++i) {
+        assertEqual(i, c2.document("test" + i).value);
+      }
+    },
+
     testDistributeShardsLikeWithDiffentKeyAttribute : function () {
-      let c1 = db._create(name1, { shardingStrategy: "community-compat", numberOfShards: 5, shardKeys: ["one"] });
```
Comment (jsteemann) [inline js/common/tests/shell/shell-community-sharding-cluster.js]: Shouldn't all the existing tests for `community-compat` sharding also exist in the future?
I see no reason for changing them to use `hash` sharding now.
I think we should add **extra** tests for the hash sharding, but the existing tests for the other sharding types should remain there too.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/557/
@jsteemann Requests addressed. Should be working now, pending the above Jenkins test.

**Pull request merged**